src/coreclr/coreutils.c

```diff
@@ -0,0 +1,371 @@
+
+        *absolutePath = strdup(realPath);
+        if (*absolutePath == NULL)
+        {
+            perror("Could not allocate buffer for path");
+            return false;
+        }
+
+        result = true;
+    }
+
+    return result;
+}
+
+bool GetDirectory(const char* absolutePath, char** directory)
+{
+    *directory = strdup(absolutePath);
+    if (*directory == NULL)
+    {
+        perror("Could not allocate buffer for path");
+        return false;
+    }
+
+    size_t len = strlen(*directory);
+    if((*directory)[len-1] == '/')
+    {
+        (*directory)[len-1] = '\0';
+        return true;
+    }
+
+    return false;
+}
+
+bool GetClrFilesAbsolutePath(const char* currentExePath, const char* clrFilesPath, char** clrFilesAbsolutePath)
+{
+    char* clrFilesRelativePath = NULL;
+    const char* clrFilesPathLocal = clrFilesPath;
+    if (clrFilesPathLocal == NULL)
+    {
+        // There was no CLR files path specified, use the folder of the current exe
+        if (!GetDirectory(currentExePath, &clrFilesRelativePath))
+        {
+            fprintf(stderr, "Failed to get directory\n");
+            return false;
+        }
+
+        clrFilesPathLocal = clrFilesRelativePath;
+
+        // TODO: consider using an env variable (if defined) as a fall-back.
+        // The windows version of the corerun uses core_root env variable
+    }
+
+    if (!GetAbsolutePath(clrFilesPathLocal, clrFilesAbsolutePath))
+    {
+        fprintf(stderr, "Failed to convert CLR files path to absolute path\n");
+        return false;
+    }
+
+    return true;
+}
+
+bool AssemblyAlreadyPresent(const char* addedAssemblies, const char* filenameWithoutExt)
+{
+    // Copy buffer as strtok munges input
+    char buf[strlen(addedAssemblies) + 1];
+    strcpy(buf, addedAssemblies);
+    const char* token = strtok(buf, ":");
+
+    while (token != NULL)
+    {
+        if (strcmp(token, filenameWithoutExt) == 0)
+        {
+            return true;
+        }
+        token = strtok(NULL, ":");
+    }
+
+    return false;
+}
+
+void AddFilesFromDirectoryToTpaList(const char* directory, char** tpaList)
+{
+    const char * const tpaExtensions[] = {
+                ".ni.dll",      // Probe for .ni.dll first so that it's preferred if ni and il coexist in the same dir
+                ".dll",
+                ".ni.exe",
+                ".exe",
+                };
+
+    DIR* dir = opendir(directory);
+    if (dir == NULL)
+    {
+        return;
+    }
+
+    // Initially empty string
+    char* addedAssemblies = malloc(1);
+    if (addedAssemblies == NULL)
+    {
+        perror("Could not allocate buffer");
+        closedir(dir);
+        return;
+    }
+
+    addedAssemblies[0] = '\0';
+
+    // Walk the directory for each extension separately so that we first get files with .ni.dll extension,
+    // then files with .dll extension, etc.
+    for (size_t extIndex = 0; extIndex < sizeof(tpaExtensions) / sizeof(tpaExtensions[0]); extIndex++)
```

Comment:
Declare the variable outside the loop to avoid a compiler warning:
```
size_t extIndex;
for (extIndex = 0; extIndex < sizeof(tpaExtensions) / sizeof(tpaExtensions[0]); extIndex++)
```